django/__init__.py

```diff
@@ -22,3 +27,38 @@ def setup(set_prefix=True):
             '/' if settings.FORCE_SCRIPT_NAME is None else settings.FORCE_SCRIPT_NAME
         )
     apps.populate(settings.INSTALLED_APPS)
+
+
+def setup(set_prefix=True):
+    """
+    This function is idempotent (calling it several times changes nothing), but
+    not reentrant (recursively calling it would have an unexpected result).
+
+    It delegates actual setup of Django to settings.DJANGO_SETUP_CALLABLE  (if set),
+    else to ``django._setup``
+
+    Returns True iff actual setup occurred.
+    """
+    global is_ready
+    if is_ready:
```

Comment:
what do you think about using a lock here as well as a global Boolean?

Reply:
I haven't thought about it, in my perspective django.setup() was always supposed to be called from single-threaded code, before request-handling threads attack, but it might be a nice protection in some cases I guess ? So I should wrap this with a non-recursive threading Lock ?

Reply:
I added thread-safety along with idempotence in https://github.com/django/django/pull/11440, thanks